subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -384,1458 +429,4003 @@ static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+	}
+
+	err = get_number(buf, &value);
+	if (err != 0) {
+		return -ENOTSUP;
+	}
+
+	switch (value) {
+	case BT_HFP_CHLD_RELEASE_ALL:
+		return chld_release_all(ag);
+	case BT_HFP_CHLD_RELEASE_ACTIVE_ACCEPT_OTHER:
+		return chld_deactivate_calls_and_accept_other_call(ag, true);
+	case BT_HFP_CALL_HOLD_ACTIVE_ACCEPT_OTHER:
+		return chld_deactivate_calls_and_accept_other_call(ag, false);
+	case BT_HFP_CALL_ACTIVE_HELD:
+		return chld_activate_held_call(ag);
+	case BT_HFP_CALL_QUITE:
+		return chld_drop_conversation(ag);
+	}
+
+	return chld_other(ag, value);
+#else
+	return -ENOEXEC;
+#endif /* CONFIG_BT_HFP_AG_3WAY_CALL */
+}
+
+static int bt_hfp_ag_bind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+{
+	uint32_t indicator;
+	uint32_t hf_indicators = 0U;
+	uint32_t supported_indicators = 0U;
+	int err;
+	char *data;
+	uint32_t len;
+
+	if (!BOTH_SUPT_FEAT(ag, BT_HFP_HF_FEATURE_HF_IND, BT_HFP_AG_FEATURE_HF_IND)) {
+		return -ENOEXEC;
+	}
+
+	if (is_char(buf, '?')) {
+		if (!is_char(buf, '\r')) {
+			return -ENOTSUP;
+		}
+
+		hfp_ag_lock(ag);
+		hf_indicators = ag->hf_indicators;
+		supported_indicators = ag->hf_indicators_of_ag & ag->hf_indicators_of_hf;
+		hfp_ag_unlock(ag);
+		len = (sizeof(supported_indicators) * 8) > HFP_HF_IND_MAX ? HFP_HF_IND_MAX
+			   : (sizeof(supported_indicators) * 8);
+		for (int i = 1; i < len; i++) {
+			bool enabled;
+
+			if (!(BIT(i) & supported_indicators)) {
+				continue;
+			}
+
+			enabled = BIT(i) & hf_indicators;
+			err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BIND:%d,%d\r\n", i,
+				enabled ? 1 : 0);
+			if (err) {
+				return err;
+			}
+
+			supported_indicators &= ~BIT(i);
+
+			if (!supported_indicators) {
+				break;
+			}
+		}
+		return 0;
+	}
+
+	if (!is_char(buf, '=')) {
+		return -ENOTSUP;
+	}
+
+	if (is_char(buf, '?')) {
+		if (!is_char(buf, '\r')) {
+			return -ENOTSUP;
+		}
+
+		data = &ag->buffer[0];
+		*data = '(';
+		data++;
+		hfp_ag_lock(ag);
+		hf_indicators = ag->hf_indicators_of_ag;
+		hfp_ag_unlock(ag);
+		len = (sizeof(hf_indicators) * 8) > HFP_HF_IND_MAX ? HFP_HF_IND_MAX
+								   : (sizeof(hf_indicators) * 8);
+		for (int i = 1; (i < len) && (hf_indicators != 0); i++) {
+			if (BIT(i) & hf_indicators) {
+				int length = snprintk(
+					data, (char *)&ag->buffer[HF_MAX_BUF_LEN - 1] - data - 2,
+					"%d", i);
+				data += length;
+				hf_indicators &= ~BIT(i);
+			}
+			if (hf_indicators != 0) {
+				*data = ',';
+				data++;
+			}
+		}
+		*data = ')';
+		data++;
+		*data = '\0';
+		data++;
+
+		err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BIND:%s\r\n", &ag->buffer[0]);
+		return err;
+	}
+
+	while (buf->len > 0) {
+		err = get_number(buf, &indicator);
+		if (err != 0) {
+			return -ENOTSUP;
+		}
+
+		if (!is_char(buf, ',')) {
+			if (!is_char(buf, '\r')) {
+				return -ENOTSUP;
+			}
+		}
+
+		if (indicator < (sizeof(hf_indicators) * 8)) {
+			hf_indicators |= BIT(indicator);
+		}
+	}
+
+	hfp_ag_lock(ag);
+	ag->hf_indicators_of_hf = hf_indicators;
+	hfp_ag_unlock(ag);
```

Comment:
I still don't quite get the locking logic here, since you're doing it on such a small scale, and for a large part of the code `ag` remains unlocked. And for the part you're doing it, it seems like this is a read and assignment of a `uint32_t` value, which will be atomic on most architectures even without locking.

Reply:
Updated.